fw/sock_srv.c

```diff
@@ -708,6 +713,30 @@ tfw_sock_srv_start_srv(TfwSrvGroup *sg, TfwServer *srv, void *hm)
 	return 0;
 }
 
+TfwSrvConn *
+tfw_sock_srv_new_steal_sk(TfwSrvConn *srv_conn)
+{
+	TfwSrvConn *conn;
+	TfwServer *srv = (TfwServer *)srv_conn->peer;
+
+	if (!(conn = tfw_sock_srv_new_conn(srv))) {
+		T_WARN_ADDR("Can't create new connection for socket stealing",
+			    &srv->addr, TFW_NO_PORT);
+		return NULL;
+	}
+
+	tfw_connection_revive((TfwConn *)conn);
+	conn->destructor = (void *)tfw_srv_conn_release;
```

Comment:
With stealing I meant exactly the call of `tfw_srv_conn_release()` for the current server connection, but we don't need to set the callback for the websocket connection `conn`.

See the `ss_linkerror()` calls in sock.c - it handles socket errors and initiates server reconnection via `tfw_srv_conn_release()`. Note that the function handles `srv_conn->sk == NULL`, so probably we can just

```C
tfw_connection_unlink_from_sk(srv_conn->sk);
conn->sk = srv_conn->sk;
srv_conn->sk = NULL;
tfw_srv_conn_release(srv_conn);
```
But be careful with other server connection users and check that we're under the socket lock just like `ss_linkerror()`.

Reply:
Ok. I use now connection destructor for reestablishing that is in essence the same as you suggest here. But i cannot see why we should unlink from socket, it would nullify all sync socket callbacks from stealed socket and prevent further socket operations.

Please re-review recent changes.